### PR TITLE
provide a way to supply additional trusted CA roots in the runtime

### PR DESCRIPTION
**Pull request by ayj**
Simplify the process of setting up multi-cluster configurations. Provide an easier way to manage multiple Root CA certs and supply them to the runtime without requiring to restart user workloads.

### Usage 
In order to supply additional Root CA cert create a ConfigMap labeled as `istio.io/extra-trust-anchors: "true"` under the Citadel CA storage namespace ("istio-system" by default). Additional Root CA PEM certificates can be supplied as values of ConfigMap data fields. Multiple ConfigMaps can be created if necessary.

Within a couple of minutes after any modification to labeled ConfigMaps the change will be noticed by Citadel and updated trusted Root CA certs bundle will be delivered to the workloads. Pilot-agent will notice updated certificates and reload Envoy within a minute or so.

Example ConfigMap:
```
apiVersion: v1
data:
  1.pem: |
    -----BEGIN CERTIFICATE-----
    MIIDBjCCAe6gAwIBAgIRAP2hlykNC0W/ozBHOmlteh0wDQYJKoZIhvcNAQELBQAw
    LDEqMCgGA1UEChMhY2x1c3Rlci5sb2NhbE1PRFlMVU1HVVFESVdXTU5QUFFJMB4X
    DTE5MDYyODA1MjE0MFoXDTI5MDYyNTA1MjE0MFowLDEqMCgGA1UEChMhY2x1c3Rl
    ci5sb2NhbE1PRFlMVU1HVVFESVdXTU5QUFFJMIIBIjANBgkqhkiG9w0BAQEFAAOC
    AQ8AMIIBCgKCAQEAxyfZ4WuBZUb0JS6r+SbLQ7VC1Xo+qvd2B2eD0nhkm1uLPltN
    3r+iW3KSEBGaRo/0MtuUv2ONyM3bGWsu6RF/eJkXLmZ7/cqmdu+7rjAvjR/Q9pBj
    XCT6gs95ihigNKy64JxRWsgLeUB7ZXHrriozaCbL3TUKlSufm5Kv7A1/KC6eMmZF
    Svt7sN/I9ORBLiRr7mSSTEyBPNFCM2mxz4iKVOcxFJNtcX5rYjBjiGehka4NRpUH
    siGS+uwz0o8TghZgccbU1DUrCa/WlBPe0wZKoqZkLWrzMLFd1GJ9i5O5+e8ZbA9n
    ISNx0/hMO6qjUhkSJi0QLPmbgOw9CYQvs0B2cwIDAQABoyMwITAOBgNVHQ8BAf8E
    BAMCAgQwDwYDVR0TAQH/BAUwAwEB/zANBgkqhkiG9w0BAQsFAAOCAQEAJRabvJ0f
    fQw6ZHTZGE+nuG/ETR4u9jIpDIsCTWkXcgEuFkMAPboGyAUc8EfVAxkV4JmjMGaQ
    dqe4v0PgLBq58fLkYrWuDW6VlHd0HPQRajNjV1M0PkgDs3cQn/0MPgHiSbDPUVv4
    PVIry+jRrQDAgrutke3JZB1O8mePRgNGkwpPA6xs4rZlmlFWseZTFiDUJo7OvDfS
    BzAH4ls/tPH9LbssFIUtWWYB43/Jexh2ZFtQ+CjrXyQW+bgqIaGH9idsXKEipxXG
    52ltCH2qVg/e8H/Zo31vr3uHs7Bpn0g/O/FDbGOi0Jl3sstBm9vfvM1sNdtWVRSW
    HsR8Qd4qPIRjBg==
    -----END CERTIFICATE-----
  2.pem: |
    -----BEGIN CERTIFICATE-----
    MIIDBTCCAe2gAwIBAgIQakpRllMLzPu+P264OTDHOjANBgkqhkiG9w0BAQsFADAs
    MSowKAYDVQQKEyFjbHVzdGVyLmxvY2FsSVZGUkNQVUtFSU5NS1hCS1ZEVkEwHhcN
    MTkwNjI4MDU0NjQ0WhcNMjkwNjI1MDU0NjQ0WjAsMSowKAYDVQQKEyFjbHVzdGVy
    LmxvY2FsSVZGUkNQVUtFSU5NS1hCS1ZEVkEwggEiMA0GCSqGSIb3DQEBAQUAA4IB
    DwAwggEKAoIBAQDNrEE2LFOMOcCxwor4vEK7iV0sS3nD9xfCVG4Qbr3AoT2QaiuW
    wPd5Fopvy6yq0n3g97w95VBl8RaeLCOWrfNua0T87XLYuX3YstN5nvsNgAGOf1El
    nQX3oSqIRMdStY/SQotO7g5n6autrayn/G++1F0/6ExiEP6hBqKslK6EY7bTGXlW
    MDVXxgedakMqoD+6ra2xkb80WwLB2F347HxP6VLGhXbjtZKpTohBFU0Sh2xUUyZJ
    l+9Tr3eyYotB/RGIleTtF1//XHdD12drlUJd+cHyR5uoLYqLmvyqe0YqPHEeFYIF
    XyJa0XaRYF+v9eyHyS4WVxmgvMsKSn67HV69AgMBAAGjIzAhMA4GA1UdDwEB/wQE
    AwICBDAPBgNVHRMBAf8EBTADAQH/MA0GCSqGSIb3DQEBCwUAA4IBAQDLFff7MNu6
    B7Mg+6cOW0IO4UiUsrc8BEweedBKlTX1jIKCYZGeEalBVcVPk9Mezaw5G49Zeki7
    rbKl88dP1+pgqd95nL1h4R4nuk/hNRSBk+JLDD0DcBbxyLMU+gBBBxyPhZ96AY6W
    06t3yJpllN9BqsASQJtwloSY7NeIlMXf1nSKWTNxwHJykRwfBrtll7hy8Kj3+ZhQ
    19HrO4qw5nnuqvAHeJfGP/7SRcrcOINrIXxMjweAk0tqtAiFD5FQb+VPrsv5HAuQ
    Qt2hPWR7ClP/SCgb3Ep99KkEzaLQ9jeSYtIdh5KhbNSSOTuNehENjgwF93QaI20i
    7sdfoObFCfSP
    -----END CERTIFICATE-----
kind: ConfigMap
metadata:
  labels:
    istio.io/extra-trust-anchors: "true"
  name: extra-ca-certs
  namespace: istio-system
```

Continuation of https://github.com/istio/istio/pull/15859 from @philrud. [Original proposal](https://docs.google.com/document/d/1SprLnDdzK_-UgTTyUGi2KhubEwpoJTRKWTC4_OP362s/edit)

### CODE REVIEW

**Comment by costinm** [comment]
Thanks for picking this up !

File: security/pkg/k8s/controller/workloadsecret.go
Comment (ayj) [inline security/pkg/k8s/controller/workloadsecret.go]: I _think_ we also want to update the `caTLSRootCert` in `istio-security` configmap to include the extra trusted roots for this to work with SDS. NodeAgent might also need to start watching for changes. @myidpt, what do you think?

Comment (ayj) [reply security/pkg/k8s/controller/workloadsecret.go]: This change would be needed if we decided to use the trust anchors for cert rotation.

File: security/pkg/pki/ca/ca.go
Comment (lei-tang) [inline security/pkg/pki/ca/ca.go]: Consider adding a comment of why using a uuid as the Org (e.g., the benefits it brings).

File: security/pkg/k8s/controller/workloadsecret.go
Comment (lei-tang) [inline security/pkg/k8s/controller/workloadsecret.go]: Consider adding comments with examples for extraTrustAnchors, e.g., explaining 'resource name', 'field name', 'field value', what exactly they are and why they are needed, and etc. 
Similarly,  consider adding comments for primaryTrustAnchor and combinedTrustAnchorsBundle.

**Comment by ayj** [comment]
added comments - ptal cc @lei-tang

**Review by lei-tang — APPROVED** [review]

**Review by howardjohn — APPROVED** [review]

File: security/pkg/k8s/controller/workloadsecret.go
Comment (incfly) [inline security/pkg/k8s/controller/workloadsecret.go]: github user name.

File: security/pkg/k8s/controller/workloadsecret.go
Comment (incfly) [reply security/pkg/k8s/controller/workloadsecret.go]: I think whatever is done for file mount has to be done for SDS. @myidpt WDTY?

File: security/pkg/k8s/controller/workloadsecret.go
Comment (incfly) [inline security/pkg/k8s/controller/workloadsecret.go]: This seems first time we introduce this terminolgoy in our code base(maybe in the doc as well?)

- Can't we just name it as `meshRootBundle` or something refers to root?
- The use case is one config map corresponds to one cluster's root? And it contains multiple root? Why we need to have multi root in a single config map? What are the key in the config map stands for?

@myidpt wdty?

**Review by incfly — COMMENTED** [review]
I don't see a flag/environment variable to guard the code change(no code changes in `istio_ca/main.go`). Are we okay with the risk?

Also I might just lack some background of the use case, so bring up some questions.

File: security/pkg/k8s/controller/workloadsecret.go
Comment (incfly) [inline security/pkg/k8s/controller/workloadsecret.go]: `security.istio.io`? I know about is not, but that's bad IMO.

File: tests/integration/security/extra_trust_anchor/extra_trust_anchor_test.go
Comment (incfly) [inline tests/integration/security/extra_trust_anchor/extra_trust_anchor_test.go]: Just want to call out: we check the secret but not the e2e envoy to envoy flow. I imagine that's what being aimed at.

**Review by incfly — COMMENTED** [review]
@JimmyCYJ fyi since you're working on improving citadel integration test without depending on k8s platform...

**Comment by incfly** [comment]
It'll be great if you can expand the PR description or somewhere of the code via comments to document the intended use case, what it enables, etc.

I grabbed a better understanding after clicking through the original PR's link.

File: security/pkg/k8s/controller/workloadsecret.go
Comment (ayj) [reply security/pkg/k8s/controller/workloadsecret.go]: I'll defer to @myidpt on naming but extra_trust_* seems reasonable to me.

User's have some flexibility in how they manage additional certs. For the multi-cluster 'join' case I'll probably use one-root-per-configmap with the configmap and key name corresponding to the name of the remote cluster. Other users might bundle multiple additional trusted roots in a single configmap. The only requirement for the key is that it is unique per configmap which is already enforced by k8s. The citadel dedups the full list of roots to minimize the bundle size distributed to workloads.

Comment (incfly) [reply security/pkg/k8s/controller/workloadsecret.go]: > For the multi-cluster 'join' case I'll probably use one-root-per-configmap with the configmap and key name corresponding to the name of the remote cluster.

Sorry I'm confused. One root per config map, but key name corresponds to the remote cluster name (we can have multiple remote cluster), how could we make it only one root?

We have multiple cluster, each of them can have different root, but this is just one level multiplicity, not sure why we need both multi config map and multi entries in each config map.

Comment (ayj) [reply security/pkg/k8s/controller/workloadsecret.go]: Users can provide one or more roots in each configmap. The configmap name and key(s) are opaque to Citadel. Only the field _value_ matters (i.e. a PEM encoded root cert). Users are free to use their own naming convention. Citadel merges the root(s) from all configmaps along it's own root to form the complete list of roots.

For multi-cluster, I'm proposing the convention be one-remote-cluster-root per configmap with the name and key both corresponding to the remote cluster's ID/name. This makes it easy for tooling to manage the list of roots by creating/deleting resources and avoids the need to patch an existing resource. Other non-multi-cluster use cases can easily co-exist with this scheme. 

multiple-roots-per-configmap isn't required but its easy to support and understand IMO, i.e. any root(s) in a configmap labeled accordingly will be appended to the list of roots. 

To restrict usage to one-root-per-configmap requires using a known key (e.g. default) which is one more config to think about and check.

Comment (incfly) [reply security/pkg/k8s/controller/workloadsecret.go]: I personally prefer not adding API/hooks to allow multi roots within a single mesh unless we have a strong use case requiring that.

Also note, based on description, we are now planning to implicitly, support `cluster.local` same trust domain and using different roots in each cluster. Is this today's single mesh multi cluster setup? Or it's something of status quo?

Just want to call out. @myidpt @smawson @nmittler

File: security/pkg/k8s/controller/workloadsecret.go
Comment (ayj) [reply security/pkg/k8s/controller/workloadsecret.go]: How extra trusted roots are added (configmap) should be orthogonal to how the resulting bundle is distributed (file vs. SDS).

File: security/pkg/k8s/controller/workloadsecret.go
Comment (ayj) [reply security/pkg/k8s/controller/workloadsecret.go]: The use case is to establish initial trust between existing clusters within minimal disruption. I expect future work will enable establishing a common CA between clusters (e.g. key rotation?). The alternative for the self-signed cert case is to copy private keys between clusters and experience downtime during rotation.

**Comment by ayj** [comment]
> I don't see a flag/environment variable to guard the code change(no code changes in istio_ca/main.go). Are we okay with the risk?

Extra roots are added unless the user explicitly creates the properly labeled configmaps. We do create a new configmap informer but that should be relatively low risk. cc @myidpt, wdyt?

**Comment by ayj** [comment]
I've addressed all comments and added an additional multi-root traffic test with envoy. This is ready for another pass. cc @incfly

File: tests/integration/security/extra_trust_anchor/extra_trust_anchor_test.go
Comment (incfly) [inline tests/integration/security/extra_trust_anchor/extra_trust_anchor_test.go]: @Monkeyanator FYI this is really interesting usage of your feature :)

File: tests/integration/security/extra_trust_anchor/main_test.go
Comment (incfly) [inline tests/integration/security/extra_trust_anchor/main_test.go]: hmm so multi trust root ca does not support the control plane mTLS? What blocks us from getting test passing with `global.ControlPlaneSecurityenabled = true`?

**Review by incfly — COMMENTED** [review]
The code itself is pretty good. Post one question regarding control plane auth.

I do want to defer the multi roots within a single trust domain cases and verification of all the combination of all use cases to @myidpt before merge.

File: tests/integration/security/extra_trust_anchor/main_test.go
Comment (ayj) [reply tests/integration/security/extra_trust_anchor/main_test.go]: This is a limitation of how and not what we're testing. The local services and remote services are sharing the same control plane (Galley/Pilot/Mixer) with different CAs. The remote envoy cannot connect to control plane with auth enabled because of the mismatch. In practice, the remote envoy would connect via a gateway or its own local control plane.

Starting another parallel in-cluster control plane required more work in the test framework. When multi-cluster support is ready in the framework we can update this test to use more than one cluster.

**Comment by ayj** [comment]
> time="18:08:53" level=debug msg="Image: \"istio-testing/mixer:istio-testing\" with ID \"sha256:8a9c9b18e12ac5ff35933393636de3758e4e861f22cb639d19b03ee1f8c34958\" not present on node \"istio-testing-control-plane\""

/test integ-security-k8s-tests_istio

File: security/pkg/k8s/controller/workloadsecret.go
Comment (myidpt) [inline security/pkg/k8s/controller/workloadsecret.go]: I think this is very expensive. We could:
1. Have a timestamp for the bundle, and in getTrustAnchorsBundle(), trigger an update if the last timestamp has passed a period, say 1 min or so.
2. Implement an listener for the bundle and update it when it's changed.
We need to think about what happens if multiple Citadels are updating the WL certs at the same time. What're the corner cases, for example, apiserver is temporarily unavailable for updating the bundle.

File: security/pkg/k8s/controller/workloadsecret.go
Comment (myidpt) [reply security/pkg/k8s/controller/workloadsecret.go]: SDS doesn't support multiple trust anchors :(
https://github.com/istio/istio/blob/master/security/pkg/server/ca/server.go#L109
https://github.com/istio/istio/blob/master/security/proto/istioca.proto#L33
The proto we are using assumes that the root is only the last cert in the cert chain, so it doesn't have an extra field to specify multiple roots / trust anchors.

Besides that, just to this question: the `caTLSRootCert` in `istio-security` configmap doesn't need to be changed. It's only needed for Node Agent to authenticate to Citadel.

Comment (ayj) [reply security/pkg/k8s/controller/workloadsecret.go]: > The caTLSRootCert in istio-security configmap doesn't need to be changed. It's only needed for Node Agent to authenticate to Citadel.

Sharing the current root CA in a configmap is very useful. Tooling doesn't need access to secrets or private keys to read the public cert.

File: security/pkg/k8s/controller/workloadsecret.go
Comment (ayj) [reply security/pkg/k8s/controller/workloadsecret.go]: I'll added a periodic timer check (option 1).

I would expect things to converge if multiple Citadel are updating certs while trust anchors are added/removed and the root CA is rotated.

File: security/pkg/k8s/controller/workloadsecret.go
Comment (myidpt) [inline security/pkg/k8s/controller/workloadsecret.go]: I don't feel this code should live here. Can we move it to configmap:
https://github.com/istio/istio/tree/master/security/pkg/k8s/configmap

Comment (ayj) [reply security/pkg/k8s/controller/workloadsecret.go]: Done. 

FWIW, I don't think its necessarily wrong for a single controller to watch multiple resources types. It did simplify the code a bit.

**Comment by ayj** [comment]
> TestAuthenticationZ 
> TestAuthenticationZ/dumps_most_recent_proxy_with_200 

/test unit-tests_istio

File: security/pkg/k8s/configmap/configmap.go
Comment (myidpt) [inline security/pkg/k8s/configmap/configmap.go]: Why this is always true?

File: security/pkg/k8s/configmap/configmap.go
Comment (myidpt) [inline security/pkg/k8s/configmap/configmap.go]: I think we should put the functions of retrieving extra trust anchors into a separate file, maybe name it as "extraTrustAnchors". The original one can be renamed as "citadelRoot".

File: security/pkg/k8s/configmap/configmap.go
Comment (myidpt) [inline security/pkg/k8s/configmap/configmap.go]: Why do you put this code here?

File: security/pkg/k8s/configmap/configmap.go
Comment (myidpt) [inline security/pkg/k8s/configmap/configmap.go]: If the CAKCB is storing all the roots, why don't we update the CAKCB in this function?

File: security/pkg/k8s/controller/workloadsecret.go
Comment (myidpt) [inline security/pkg/k8s/controller/workloadsecret.go]: Can we use the root cert in CAKCB as the source of truth for multiple trust anchors? It should be kept in-sync with the K8s ConfigMap by the configmap.controller.

File: security/pkg/k8s/controller/workloadsecret.go
Comment (ayj) [reply security/pkg/k8s/controller/workloadsecret.go]: Probably not. This code path became less clear to me after the cert rotation was added and I had to rebase my changes a few times.

**Comment by howardjohn** [comment]
@ayj is this still needed? looks like it is still waiting for securiyt approval

**Comment by ayj** [comment]
I'll see if someone on the security team wants to take this up.

**Pull request closed without merging**
